Fix bin edge/centre bugs PDFFourierTransform2

RichardWaiteSTFC opened the pull request:
### Description of work

Fix bin edge/centre bugs PDFFourierTransform2

The definite integral here (and elsewhere) should be evaluated at the bin edges.
https://github.com/mantidproject/mantid/blob/c9e2567e14cac63ade5abc192bd076a9b5077307/Framework/Algorithms/src/PDFFourierTransform2.cpp#L511

The lorch filter here  requires the the Q at the bin center
https://github.com/mantidproject/mantid/blob/c9e2567e14cac63ade5abc192bd076a9b5077307/Framework/Algorithms/src/PDFFourierTransform2.cpp#L524

The first bin should be centred on bin-width/2
https://github.com/mantidproject/mantid/blob/c9e2567e14cac63ade5abc192bd076a9b5077307/Framework/Algorithms/src/PDFFourierTransform2.cpp#L487

Fixes #38961  -though the oscillations are mostly due to S(Q)-1 not tending to 0 at high Q

**Report to:** POLARIS

### To test:

(1) Run this script with this file
[s_of_q_minus_one_merged.nxs.txt](https://github.com/user-attachments/files/19755787/s_of_q_minus_one_merged.nxs.txt)


```
# import mantid algorithms, numpy and matplotlib

from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from pystog import Transformer, FourierFilter

ws_sqm1_merged = LoadNexus(r"C:/Users/xhg73778/Documents/POLARIS/s_of_q_minus_one_merged.nxs")

deltar = None
do_lorch = True

ws_gr_merged = PDFFourierTransform(InputWorkspace=ws_sqm1_merged, SofQType='S(Q)-1', PDFType='g(r)', 
                                   DeltaR=deltar, RMax=None, Filter=do_lorch)                 
ws_sqm1_merged_pts = ConvertToPointData(InputWorkspace=ws_sqm1_merged, OutputWorkspace=ws_sqm1_merged.name() + "_pts")
ws_gr_merged_pts = PDFFourierTransform(InputWorkspace=ws_sqm1_merged_pts, SofQType='S(Q)-1', PDFType='g(r)', 
                                   DeltaR=deltar, RMax=None, Filter=do_lorch)
                                   
qedge = ws_sqm1_merged.readX(0)[:,None]
dq = qedge[1:] - qedge[:-1]  # np.diff messes with array sahpe and broadcasting
qcens = 0.5*(qedge[1:] + qedge[:-1])
sqm1 = ws_sqm1_merged.readY(0)[:,None]
rcens = ws_gr_merged.readX(0)
rcens = 0.5*(rcens[1:] + rcens[:-1])
r, gr, dgr = Transformer().F_to_g(qcens, qcens*sqm1, rcens, rho=0.0155, lorch=do_lorch)

fig, axes = plt.subplots(dpi=156.0, edgecolor='#ffffff', num='ws_gr_merged-1', subplot_kw={'projection': 'mantid'})
axes.plot(ws_gr_merged, color='#1f77b4', label='ws_gr_merged: spec 1', wkspIndex=0)
axes.plot(ws_gr_merged_pts, color='#ff7f0e', label='ws_gr_merged_pts: spec 1', linestyle='--', wkspIndex=0, zorder=2.1)
axes.plot(r, gr, color="k", ls='--', marker='o', markersize=2, label='pystog (F_to_g)')
axes.set_title('ws_gr_merged')
axes.set_xlabel('Atomic Distance ($\\AA$)')
axes.set_ylabel('PDF')
axes.set_xlim([0.5, 7])
axes.set_ylim([-700, 900])
legend = axes.legend(fontsize=8.0).set_draggable(True).legend
fig.show()
```

This produces the figure on the RHS
<img width="646" alt="image" src="https://github.com/user-attachments/assets/009c4879-c6a9-4162-bc07-93ec8bead7f1" />

It can be seen that the peak center now agrees with the g(r) produced by pystog (library that performs fourier transform differently so is a good check). Previously the oscillations in the pystog g(r) on the LHS figure is due to the incorrect bin centers coming from PDFFourierTransform. If you set `DeltaR=0.02` you over-sample g(r) and you see these oscillations in both pystog and Mantid

<img width="313" alt="image" src="https://github.com/user-attachments/assets/4bcf0f47-0f6b-443e-a1ef-066c1d9f9c4e" />

(2) Run the backward transform back to Q (with `do_lorch=True`)
```
# convert back to Q
ws_sqm1_bk = PDFFourierTransform(InputWorkspace=ws_gr_merged,             
            SofQType="S(Q)-1",
            PDFType="g(r)",
            QMin=qcens.flat[0],
            Qmax=qcens.flat[-1],
            deltaQ=dq.flat[0],
            Direction="Backward", 
            Filter=do_lorch)
q_bk = ws_sqm1_bk.readX(0)
q_bk = 0.5*(q_bk[:-1] + q_bk[1:])
q_bk, qsqm1, dqsqm1_bk = Transformer().g_to_F(rcens, ws_gr_merged.readY(0), q_bk, rho=0.0155, lorch=do_lorch)

fig, axes = plt.subplots(subplot_kw={'projection': 'mantid'})
axes.plot(ws_sqm1_bk, color='#1f77b4', label='ws_gr_merged: spec 1', wkspIndex=0)
axes.plot(q_bk, qsqm1/q_bk, color="k", ls='--', marker='o', markersize=2, label='pystog (F_to_g)')
legend = axes.legend(fontsize=8.0).set_draggable(True).legend
fig.show()
```
This produces the figure on the RHS

<img width="584" alt="image" src="https://github.com/user-attachments/assets/9efde6bd-dfa0-4df7-aaed-45fab719a2ff" />

The match is a lot better, the integral is done slightly differently in the two programs and the back transform of transformed data always has artefacts etc.

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

RichardWaiteSTFC committed:
Fix bin edge/centre bugs in PDFFourierTransform2

RichardWaiteSTFC committed:
Fix tests - small changes mostly due to x shifted by half a bin

RichardWaiteSTFC committed:
Add release note

RichardWaiteSTFC committed:
Fix tests and add test to assert points and hist input give same out

sf1919 commented:
As this is failing tests and further commit(s) are expected we won't re-run any tests on the new Linux nodes.

RichardWaiteSTFC committed:
Update peak position in system test and adjust pdf values

Peak is actually at ~3.85 (as confirmed by ADDIE simulation)
Was thought to be at 3.9 because of the out by a bin bug.
Updated pdf values very similar.

RichardWaiteSTFC commented:
To fix systems test needed to adjust the r value at which the test asserts the pdf value - the peak for silicon is actually at r~3.85 Ang as confirmed by ADDIE simulation - was set to 3.9 Ang in test probably due to the out by bin error fixed here
![image](https://github.com/user-attachments/assets/491f35d4-aab0-4638-9c45-e3cc9d318cd8)

RichardWaiteSTFC committed:
Fix doc test

warunawickramasingha reviewed (approved):
Code changes looks good and I have followed the given test steps to get the desired outputs. Happy to approve!

jclarkeSTFC reviewed (approved):
Code changes look good, wiggles have been removed from the output.
